spark/sql-13/src/main/scala/org/elasticsearch/spark/sql/SchemaUtils.scala

```diff
@@ -273,16 +273,21 @@ private[sql] object SchemaUtils {
     val rowInfo = (new Properties, new Properties)
 
     doDetectInfo(rowInfo, ROOT_LEVEL_NAME, struct)
-    val csv = SettingsUtils.determineSourceFields(settings)
-    // if a projection is applied (filtering or projection) use that instead
-    if (StringUtils.hasText(csv)) {
-      if (settings.getReadMetadata) {
-        rowInfo._1.setProperty(ROOT_LEVEL_NAME, csv + StringUtils.DEFAULT_DELIMITER + settings.getReadMetadataField)
-      }
-      else {
-        rowInfo._1.setProperty(ROOT_LEVEL_NAME, csv)
-      }
-    }
+
+    val requiredFields = settings.getProperty(Utils.DATA_SOURCE_REQUIRED_COLUMNS)
+    val sourceFields = SettingsUtils.determineSourceFields(settings)
+
+    // In case when user selected specific fields (we want to keep the same order in a spark sql row)
+    if (StringUtils.hasText(requiredFields))
+      rowInfo._1.setProperty(ROOT_LEVEL_NAME, requiredFields)
+
+    // In case when we read all fields including metadata
+    else if (StringUtils.hasText(sourceFields) && settings.getReadMetadata)
+      rowInfo._1.setProperty(ROOT_LEVEL_NAME, sourceFields + StringUtils.DEFAULT_DELIMITER + settings.getReadMetadataField)
```

Comment:
Yes, you are right, one condition is enough. My original assumption was:
```
if (StringUtils.hasText(requiredFields)) {
   // for queries like dataFrame.select("specific", "fields").take(1)
} else if (StringUtils.hasText(sourceFields) && settings.getReadMetadata) {
   // for queries like dataFrame.take(1) when "es.read.metadata" is set to true
} else if (StringUtils.hasText(sourceFields) {
   // for queries like dataFrame.take(1) when "es.read.metadata" is set to false
}
```
But it isn't the case, because of the way how we set `requiredFields`.